File: Problems/100_Same_Tree/Same_Tree.py
# coding: utf-8

import sys
import time

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def isSameTree(self, p, q):
        """
        :type p: TreeNode
        :type q: TreeNode
        :rtype: bool
        """
        if (p == None) and (q == None):
            return True
        elif (p == None) or (q == None):
            return False
        if p.val == q.val:
            if self.isSameTree(p.left, q.left):
                return self.isSameTree(p.right, q.right)
            else:
                return False
        else:
            return False


def set_node(data):
    if data == None:
        return
    new_node = TreeNode(data[0])
    if data[1] != "null":
        new_node.left = TreeNode(data[1])
    if data[2] != "null":
        new_node.right = TreeNode(data[2])
    return new_node


def output_node(node):
    tempStr = ""
    if node != None:
        tempStr += node.val + ","
    if node.left != None:
        tempStr += node.left.val + ","
    else:
        tempStr += "null,"
    if node.right != None:
        tempStr += node.right.val
    else:
        tempStr += "null"

    return tempStr


def main():
    args = sys.argv
    argc = len(args)

    print("args[0] = %s args[1] = %s" %(args[0], args[1]) )
    flds = args[1].rstrip().split(chr(0x09)) # remove LF and splip [TAB]
    print("flds[0] = %s fls[1] = %s" %(flds[0], flds[1]) )
    arg1 = flds[0].split(',')
    arg2 = flds[1].split(',')

    p = set_node(arg1)
    q = set_node(arg2)

    print("node p = %s" %(output_node(p)))
    print("node q = %s" %(output_node(q)))

    time0 = time.time()

    sl = Solution()
    print(sl.isSameTree(p, q))

    time1 = time.time()
    print("Execute time ... : %f[s]" %(time1 - time0))
    print()


if __name__ == "__main__":
    main()

File: Problems/0100_Same_Tree/Same_Tree.cs
using System;

// Definition for a binary tree node.
 public class TreeNode {
    public int val;
    public TreeNode left;
    public TreeNode right;
    public TreeNode(int x) { val = x; }
}

public class Solution
{
    public bool IsSameTree(TreeNode p, TreeNode q)
    {
        if (p == null && q == null) {
            return true;
        }
        else if (p == null || q == null)
        {
            return false;
        }

        if (p.val == q.val) {
            if (IsSameTree(p.left, q.left)) {
                return IsSameTree(p.right, q.right);
            }
            else {
                return false;
            }
        }
        else {
            return false;
        }
    }

    private TreeNode set_node(string[] data)
    {
        if (data.Length == 0) {
            return null;
        }

        TreeNode new_node = new TreeNode(int.Parse(data[0]));
        if (data[1] != "null") {
            new_node.left = new TreeNode(int.Parse(data[1]));
        }
        if (data[2] != "null") {
            new_node.right = new TreeNode(int.Parse(data[2]));
        }

        return new_node;
    }

    private string output_node(TreeNode node)
    {
        string tempStr = "";

        if (node != null) {
            tempStr += node.val + ",";
        }
        if (node.left != null) {
            tempStr += node.left.val + ",";
        }
        else {
            tempStr += "null,";
        }
        if (node.right != null) {
            tempStr += node.right.val;
        }
        else
        {
            tempStr += "null";
        }

        return tempStr;
    }
    public void Main(string args)
    {
        Console.WriteLine("args = " + args );

    //    string[] workStr = args.Split('\t');
        string[] workStr = args.Split((char)0x09);    // [TAB]

        string[] arg1 = workStr[0].Split(',');
        string[] arg2 = workStr[1].Split(',');

        TreeNode p = set_node(arg1);
        TreeNode q = set_node(arg2);

        Console.WriteLine("p = " + output_node(p));
        Console.WriteLine("q = " + output_node(q));

        System.Diagnostics.Stopwatch sw = new System.Diagnostics.Stopwatch();
        sw.Start();

        Console.WriteLine("result = " + IsSameTree(p, q).ToString() );

        sw.Stop();
        Console.WriteLine("Execute time ... " + sw.ElapsedMilliseconds.ToString() + "ms");
    }
}
